File: VB/WinWebSolution.Module/BO.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.Xpo
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Base

Namespace WinWebSolution.Module
    <DefaultClassOptions> _
    Public Class Producer
        Inherits BaseObject

        Public Sub New(ByVal s As Session)
            MyBase.New(s)
        End Sub

        Public Property Name() As String
            Get
                Return GetPropertyValue(Of String)("Name")
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Name", value)
            End Set
        End Property

        <Aggregated, Association("Producer-Transactions")> _
        Public ReadOnly Property Transactions() As XPCollection(Of Transaction)
            Get
                Return GetCollection(Of Transaction)("Transactions")
            End Get
        End Property
    End Class

    <DefaultClassOptions> _
    Public Class Consumer
        Inherits BaseObject

        Public Sub New(ByVal s As Session)
            MyBase.New(s)
        End Sub

        Public Property Name() As String
            Get
                Return GetPropertyValue(Of String)("Name")
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Name", value)
            End Set
        End Property

        <Aggregated, Association("Consumer-Transactions")> _
        Public ReadOnly Property Transactions() As XPCollection(Of Transaction)
            Get
                Return GetCollection(Of Transaction)("Transactions")
            End Get
        End Property
    End Class

    Public Class Transaction
        Inherits BaseObject

        Public Sub New(ByVal s As Session)
            MyBase.New(s)
        End Sub

        Public Property Product() As String
            Get
                Return GetPropertyValue(Of String)("Product")
            End Get
            Set(ByVal value As String)
                SetPropertyValue(Of String)("Product", value)
            End Set
        End Property

        Public Property Amount() As Decimal
            Get
                Return GetPropertyValue(Of Decimal)("Amount")
            End Get
            Set(ByVal value As Decimal)
                SetPropertyValue(Of Decimal)("Amount", value)
            End Set
        End Property

        <Association("Producer-Transactions")> _
        Public Property Producer() As Producer
            Get
                Return GetPropertyValue(Of Producer)("Producer")
            End Get
            Set(ByVal value As Producer)
                SetPropertyValue(Of Producer)("Producer", value)
            End Set
        End Property

        <Association("Consumer-Transactions")> _
        Public Property Consumer() As Consumer
            Get
                Return GetPropertyValue(Of Consumer)("Consumer")
            End Get
            Set(ByVal value As Consumer)
                SetPropertyValue(Of Consumer)("Consumer", value)
            End Set
        End Property
    End Class
End Namespace

File: CS/WinWebSolution.Module/BO.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.Xpo;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Base;

namespace WinWebSolution.Module {
    [DefaultClassOptions]
    public class Producer : BaseObject {
        public Producer(Session s) : base(s) { }

        public string Name;

        [Aggregated]
        [Association("Producer-Transactions")]
        public XPCollection<Transaction> Transactions {
            get { return GetCollection<Transaction>("Transactions"); }
        }
    }

    [DefaultClassOptions]
    public class Consumer : BaseObject {
        public Consumer(Session s) : base(s) { }

        public string Name;

        [Aggregated]
        [Association("Consumer-Transactions")]
        public XPCollection<Transaction> Transactions {
            get { return GetCollection<Transaction>("Transactions"); }
        }
    }

    public class Transaction : BaseObject {
        public Transaction(Session s) : base(s) { }

        public string Product;
        public decimal Amount;

        private Producer producer;
        [Association("Producer-Transactions")]
        public Producer Producer {
            get { return producer; }
            set { SetPropertyValue("Producer", ref producer, value); }
        }

        private Consumer consumer;
        [Association("Consumer-Transactions")]
        public Consumer Consumer {
            get { return consumer; }
            set { SetPropertyValue("Consumer", ref consumer, value); }
        }
    }
}
